deepcopy: refactor output of T.Copy() T*

stevvooe opened the pull request:
To make the deepcopy plugin interact better with externally defined
    types, it has been refactored to define a more flexible interface, based
    around `CopyFrom`. `CopyFrom` takes a `src` argument and copies `src`
    the fields of the method receiver. On top of this, we can build
    `deepcopy.Copy` which takes a pre-allocated `dst` and `src` argument.
    Because we provide our own copy implementation, as a parameter of the
    plugin, we can inject our own implementations in the copy function
    without affecting valid compilation of generated code.

    In particular, this will allow us to use externally defined timestamp
    and duration types, in addition to other well known types, that were
    previously unusable due to the requirement that the Copy() function be
    generated on types.

    To preserve existing direct usage of `T.Copy() *T`, we still generated
    the method for use in code. Since this code is typically for direct
    usage, and has access to the Go compiler when under use (as opposed to
    the protoc generation stack), this usage is safe and encouraged.

    Signed-off-by: Stephen J Day <stephen.day@docker.com>

aluzzardi commented:
Hold - I think this will make cherry picking a pain during the freeze. Do you agree?

stevvooe commented:
> Hold - I think this will make cherry picking a pain during the freeze. Do you agree?

Possibly.

I may place this in a larger PR that we can merge after freeze.

aluzzardi commented:
Thanks, appreciated

stevvooe commented:
Updated to include deepcopy refactor and regenerated protobufs.

aaronlehmann commented:
Thanks a lot for doing this.

To provide some context, #1761 will be able to take advantage of this to use the `Any` type.

Note that CI is showing a lint error.

stevvooe commented:
@aaronlehmann This should enable the following:

1. Removal of forked timestamp/duration protobuf definitions.
2. We can now use the `stdtime` plugin from gogo protobuf, which uses native `time.Time`/`time.Duration`.
3. We can now move to use the `Any` type in a lot of cases. This will be great opportunity to introduce plugins and extensions to swarmkit. Imagine `TaskSpec.Runtime.Any`.
4. The `deepcopy` plugin can more easily be contributed upstream.

There may be a few more benefits that can be achieved with some further changes.

When the merge window opens up again, I'll get this into shape and start tackling the first two, at least.

stevvooe commented:
@aaronlehmann @aluzzardi Are we ready to rebase and merge this?

aaronlehmann commented:
I think @aluzzardi wants to wait until the 1.13 release to merge it, but I'd be thrilled to see this ready to go.

aaronlehmann commented:
@stevvooe: I think we can move forward with this now. Do you have the time to complete it? If you'd like I can carry it.

stevvooe commented:
@aaronlehmann Rebased. Fairly painless.

Please doublecheck that the dependency changes are actually updates. That was a little tricky, so we might have to leave those commits alone and then move forward from there.

aaronlehmann commented:
CI appears to have hung, and I'm seeing a test failure when I run locally:

```
--- FAIL: TestUpdaterRollback (1.16s)
        Error Trace:    update_test.go:219
	Error:		Not equal: "image1" (expected)
			        != "image2" (actual)

        Error Trace:    update_test.go:223
	Error:		Not equal: "image1" (expected)
			        != "image2" (actual)

        Error Trace:    update_test.go:242
	Error:		Not equal: "image2" (expected)
			        != "image1" (actual)

        Error Trace:    update_test.go:246
	Error:		Not equal: "image2" (expected)
			        != "image1" (actual)

        Error Trace:    update_test.go:250
	Error:		Not equal: "image2" (expected)
			        != "image1" (actual)
```

aaronlehmann commented on protobuf/plugin/deepcopy/deepcopy.go:
```diff
@@ -22,201 +22,250 @@ func (d *deepCopyGen) Name() string {
-	oneOfFuncs := make(map[string][]func())
+	d.P("func (m *", ccTypeName, ") CopyFrom(src interface{})", " {")
+	d.In()
+	d.P("if m == nil && src == nil {")
```
Should this be `||`?

stevvooe replied:
Calling `CopyFrom` on a nil object should probably panic. If `src` is nil, we should probably clear the object.

aaronlehmann commented:
Ah, the problem was fairly obvious, but it took me awhile to see it.

```go
func (m *EndpointSpec) Copy() *EndpointSpec {
        o := &EndpointSpec{}
        o.CopyFrom(m)
        return o
}
```

If `m == nil`, this must return `nil`. Otherwise, we end up replacing nil pointers with pointers to empty structs when making deep copies, and this causes `reflect.DeepEqual` to flag copied objects as different from the source.

This appears to fix the problem:

```go
diff --git a/protobuf/plugin/deepcopy/deepcopy.go b/protobuf/plugin/deepcopy/deepcopy.go
index 2594a76..d1014d7 100644
--- a/protobuf/plugin/deepcopy/deepcopy.go
+++ b/protobuf/plugin/deepcopy/deepcopy.go
@@ -35,6 +35,11 @@ func (d *deepCopyGen) genMsgDeepCopy(m *generator.Descriptor) {
        // Generate backwards compatible, type-safe Copy() function.
        d.P("func (m *", ccTypeName, ") Copy() *", ccTypeName, "{")
        d.In()
+       d.P("if m == nil {")
+       d.In()
+       d.P("return nil")
+       d.Out()
+       d.P("}")
        d.P("o := &", ccTypeName, "{}")
        d.P("o.CopyFrom(m)")
        d.P("return o")
```

We should probably also add a unit test.

stevvooe commented:
Made the following changes:

- Removed `nil` kid gloves in `CopyFrom` and allow that interface to panic with nil args or receiver. In either case, there isn't a "correct" behavior.
- If receiver in `T.Copy()` is `nil`, return `nil`
- Added test case to ensure `T(nil).Copy()` returns `nil`

aaronlehmann commented:
Thanks.

In the generated code for `CopyFrom` I see stuff like:

```
	if o.Endpoint != nil {
		m.Endpoint = &EndpointSpec{}
		github_com_docker_swarmkit_api_deepcopy.Copy(m.Endpoint, o.Endpoint)
	}
```

Should there be an `else` branch that sets `m.Endpoint = nil` otherwise? I don't think this will make any difference with the current code, which only uses `CopyFrom` into an empty object. But if anything ever tries to use `CopyFrom` with a destination that already has fields initialized, it looks like some of those fields may be skipped if they are nil in the source object.

stevvooe commented:
> Should there be an else branch that sets m.Endpoint = nil otherwise?

Yes.

aaronlehmann commented:
LGTM

Let's fix `CopyFrom` into dirty objects in a followup PR.

stevvooe commented:
@aaronlehmann Let's go ahead and merge this and I'll send that optimization in second PR.

aaronlehmann commented:
Oh dear - I think merging this my have introduced a problem :(. There was a raft test failure that we saw in CI for this PR and dismissed as a fluke when it didn't reoccur, but I just saw it again on master after this was merged. I think updating protobuf or grpc has caused problems with raft. Maybe we should try updating the vendored raft code (#1862)?

aaronlehmann commented:
cc @LK4D4

stevvooe commented:
@aaronlehmann Did we get to the bottom of the problem described in https://github.com/docker/swarmkit/pull/1763#issuecomment-274223685?

aaronlehmann commented:
> @aaronlehmann Did we get to the bottom of the problem described in #1763 (comment)?
> 

Updating raft seems to have helped.
